File: opencv_test2/osama_part/GP Android OpenCV - Gesture Recognition/src/edu/alexeng/greco/Trainner.java
package edu.alexeng.greco;

import java.io.File;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfInt;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Point;
import org.opencv.core.Rect;
import org.opencv.core.RotatedRect;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.core.TermCriteria;
import org.opencv.highgui.Highgui;
import org.opencv.imgproc.Imgproc;
import org.opencv.ml.CvSVM;
import org.opencv.ml.CvSVMParams;

import HelperClasses.FileReader;
import HelperClasses.FileWriter;
import android.os.Environment;
import android.util.Log;

public class Trainner {

	private List<MatOfPoint> contours;
	private RotatedRect handBoundary;
	private int largestContourIndex;
	private Mat hand, temp;
	private Mat frameInProcess;
	private MatOfPoint largestContour;
	private final int gridSize = 10;
	private final int dataSetSize = 177;
	private float[][] data;
	private float[] labels;
	private CvSVM SVM;
	private int row;

	private FileReader reader;

	public Trainner() {
		// some ini
		contours = new ArrayList<MatOfPoint>();
		handBoundary = new RotatedRect();
		largestContourIndex = -1;
		frameInProcess = new Mat();
		data = new float[dataSetSize][100];
		labels = new float[dataSetSize];
		boolean loop = false;
		row = 0;
		reader = new FileReader();
		ArrayList<String> paths = reader.getImgs();
		String root = Environment.getExternalStorageDirectory().getPath()
				+ "//";
		for (int i = 0; i < (paths.size() - 1); i++) {
			String path = root + paths.get(i);
			Mat img = Highgui.imread(path);
			process(img);
		}
		for (int j = 0; j < 100; j++)
			data[row][j] = 0;

		train();
		FileWriter writer = new FileWriter(data);
		Log.d("hamada", "finishhhhhhhhhh");

	}

	public void process(Mat mFrame) {
		Imgproc.GaussianBlur(mFrame, frameInProcess, new Size(5, 5), 0, 0);
		// Imgproc.erode(frameInProcess, frameInProcess, new Mat());
		// Imgproc.dilate(frameInProcess, frameInProcess, new Mat());
		Imgproc.cvtColor(frameInProcess, frameInProcess,
				Imgproc.COLOR_BGR2YCrCb);
		Core.inRange(frameInProcess, new Scalar(0, 131, 80), new Scalar(255,
				185, 135), frameInProcess);
		Imgproc.medianBlur(frameInProcess, frameInProcess, 5);
		// Core.bitwise_not(frameInProcess, frameInProcess);
		hand = frameInProcess.clone();
		// finding contours
		Imgproc.findContours(frameInProcess, contours, new Mat(),
				Imgproc.RETR_LIST, Imgproc.CHAIN_APPROX_SIMPLE);

		// finding largest contour
		largestContourIndex = -1;
		largestContour = null;
		double maxArea = 0;
		Iterator<MatOfPoint> each = contours.iterator();
		for (int i = 0; each.hasNext(); i++) {
			MatOfPoint wrapper = each.next();
			double area = Imgproc.contourArea(wrapper);
			if (area > maxArea) {
				largestContourIndex = i;
				largestContour = wrapper;
				maxArea = area;
			}
		}
		if (largestContourIndex == -1 || largestContour == null
				|| largestContour.total() == 0)
			return;

		handBoundary = Imgproc.minAreaRect(new MatOfPoint2f(largestContour
				.toArray()));
		// copy the hand in its own image
		double max = handBoundary.size.width;
		if (max < handBoundary.size.height)
			max = handBoundary.size.height;
		double left = handBoundary.center.x - (max / 2) - 25, top = handBoundary.center.y
				- (max / 2) - 25;
		double width = max + 50, hight = max + 50;
		if (left < 0)
			left = 0;
		if (top < 0)
			top = 0;
		if ((left + width) > hand.width())
			width = hand.width() - left;
		if ((top + hight) > hand.height())
			hight = hand.height() - top;
		if (width <= 0 || hight <= 0) {
			left = 0;
			top = 0;
			width = hand.width();
			hight = hand.height();
		}
		Rect rounded = new Rect(new double[] { left, top, width, hight });

		hand = hand.submat(rounded);

		extractFeature();

		row++;
	}

	private void extractFeature() {

		int unitWidth = hand.cols() / gridSize;
		int unitHeight = hand.rows() / gridSize;
		int index = 0;
		for (int i = 0; i < gridSize; i++) {
			for (int j = 0; j < gridSize; j++) {
				Mat subImg = new Mat(hand, new Rect(new double[] {
						j * unitWidth, i * unitHeight, unitWidth, unitHeight }));
				int white = Core.countNonZero(subImg);
				if (white > 0)
					data[row][index] = 0;
				else
					data[row][index] = 1;
				index++;
				subImg.release();
			}
		}
	}

	private void train() {

		CvSVMParams params = new CvSVMParams();
		params.set_svm_type(CvSVM.C_SVC);
		params.set_kernel_type(CvSVM.LINEAR);
		params.set_term_crit(new TermCriteria(TermCriteria.MAX_ITER, 100, 1e-6));

		assign_lables();

		Mat labelsMat = new Mat(dataSetSize, 1, CvType.CV_32FC1);
		labelsMat.put(0, 0, labels);

		Mat trainingMat = new Mat(dataSetSize, 100, CvType.CV_32FC1);
		for (int i = 0; i < dataSetSize; i++)
			trainingMat.put(i, 0, data[i]);

		SVM = new CvSVM();
		SVM.train(trainingMat, labelsMat, new Mat(), new Mat(), params);
		File datasetFile = new File(Environment.getExternalStorageDirectory(),
				"AndroidDataset.xml");
		SVM.save(datasetFile.getAbsolutePath());
	}

	private void assign_lables() {
		int index = 0;
		// up
		for (int i = 0; i < 14; i++) {
			labels[index] = 1;
			index++;
		}
		// open
		for (int i = 0; i < 22; i++) {
			labels[index] = 2;
			index++;
		}
		// capture
		for (int i = 0; i < 38; i++) {
			labels[index] = 3;
			index++;
		}
		// call
		for (int i = 0; i < 25; i++) {
			labels[index] = 4;
			index++;
		}
		// left
		for (int i = 0; i < 22; i++) {
			labels[index] = 5;
			index++;
		}
		// right
		for (int i = 0; i < 22; i++) {
			labels[index] = 6;
			index++;
		}
		// closed
		for (int i = 0; i < 15; i++) {
			labels[index] = 7;
			index++;
		}
		// start
		for (int i = 0; i < 18; i++) {
			labels[index] = 8;
			index++;
		}
		// false
		for (int i = 0; i < 1; i++) {
			labels[index] = 9;
			index++;
		}
	}

}

File: Gesture Recognition Project/Gridding Approach/trainner.cpp
/*
 * trainner.cpp
 *
 *  Created on: Mar 23, 2014
 *      Author: Mohamed
 */

#include "trainner.h"

trainner::trainner() {

	/*
	 * reading images, convert them to binary,extract features
	 */
	FileReader reader("pathsF.txt");
	FileWriter writer("out.txt");
	string line = "";
	IplImage* img;
	IplImage*gray_im;
	IplImage*gray_img;
	CvSeq* contour; //pointer to a contour.
	CvMemStorage* space = cvCreateMemStorage(0);
	CvBox2D hand_boundary;
	CvSeq* largest_contour;
	int i = 0;
	int all_time[dataset_size];
	while ((line = reader.readFile()) != "") {
		std::clock_t start;

		start = std::clock();

		//load the img
		img = cvLoadImage(line.c_str());
		cvSmooth(img, img, CV_GAUSSIAN, 5, 5);
		gray_im = cvCloneImage(img);
		cvCvtColor(img, gray_im, CV_BGR2YCrCb);
		gray_img = cvCreateImage(cvGetSize(gray_im), 8, 1);
		cvInRangeS(gray_im, cvScalar(0, 131, 80), cvScalar(255, 185, 135),
				gray_img);
		cvSmooth(gray_img, gray_img, CV_MEDIAN, 5, 5);

		hand = cvCloneImage(gray_img);
		//finding all contours in the image
		cvFindContours(gray_img, space, &contour, sizeof(CvContour),
				CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE, cvPoint(0, 0));
		//iterate through each contour
		double max_area = 0;
		double area = 0;
		//finding largest contour
		while (contour) {
			area = cvContourArea(contour);
			if (area > max_area) {
				max_area = area;
				largest_contour = contour;
			}
			contour = contour->h_next;
		}

		if (largest_contour && largest_contour->total > 0) {
			hand_boundary = cvMinAreaRect2(largest_contour);

			float max = hand_boundary.size.width;
			if (max < hand_boundary.size.height)
				max = hand_boundary.size.height;
			//copy the hand in its own image
			CvRect rounded =
					cvRect(hand_boundary.center.x - (max / 2) - 25,
							hand_boundary.center.y - (max / 2) - 25, max + 50,
							max + 50);
			cvSetImageROI(hand, rounded);
			hand = cvCloneImage(hand);
//			cvShowImage("image", hand);
			cvWaitKey(0);
			cvReleaseImage(&gray_img);
			cvClearSeq(largest_contour);
			//			string bin = extract_feature();
			//write to file
			//			writer.writeFile(bin);
			extract_feature(i);

		} else {
			for (int j = 0; j < number_of_features; j++)
				data[i][j] = 0.0;
		}
		int timm = (std::clock() - start) / (double) (CLOCKS_PER_SEC / 1000);
		all_time[i] = timm;
		i++;

	}
	int sum = 0;

	for (int i = 0; i < dataset_size; i++) {
		sum += all_time[i];
	}
	sum = sum / dataset_size;
	cout << sum << endl;
	reader.~FileReader();

	// now train the classifier
	train();

	//print features
	for (int i = 0; i < dataset_size; i++) {
		ostringstream oss;
		if (i < 11)
			oss << "up";
		else if (i < 30)
			oss << "open";
		else if (i < 60)
			oss << "capture";
		else if (i < 83)
			oss << "call";
		else if (i < 101)
			oss << "left";
		else if (i < 125)
			oss << "right";
		else if (i < 136)
			oss << "closed";
		else if (i < 149)
			oss << "start";
		else if (i < 159)
			oss << "Lup";
		else if (i < 173)
			oss << "Lopen";
		else if (i < 190)
			oss << "Lcapture";
		else if (i < 197)
			oss << "Lcall";
		oss << ",";
		for (int j = 0; j < number_of_features; j++) {
			if (data[i][j] == 0.0)
				oss << "0";
			else
				oss << "1";
			oss << ",";
		}
		string name = oss.str();
		writer.writeFile(name);
	}
	writer.~FileWriter();
}
void trainner::extract_feature(int row) {
//	stringstream oss;
//	oss<<row<<"_";
//	if (row >= 11 && row <30){
//		oss<<"openBinary.jpg";
//		cvSaveImage(oss.str().c_str(), hand);
//	}
	Mat img_data(hand);
	int unitWidth = img_data.cols / grid_size; // you had image.rows / n;
	int unitHeight = img_data.rows / grid_size;
	Mat dctImage = img_data.clone();
	String binary = "";
	int index = 0;
	for (int i = 0; i < grid_size; i++) { //i is row index
		// inner loop added so that more than one row of tiles written
		for (int j = 0; j < grid_size; j++) { // j is col index
			Mat subImage = dctImage(
					Rect(j * unitWidth, i * unitHeight, unitWidth, unitHeight));
			//convert it to gray then binary
			int white = countNonZero(subImage);
			//			cout<<"white : "<<white<<endl;
			if (white > 0)
				data[row][index] = 0.0;
			//				binary.append("0,");
			else
				data[row][index] = 1.0;
			//				binary.append("1,");

			index++;
		}
		//		binary.append(" ");
	}
	//	return binary;
	//	return features;
}
void trainner::train() {
	//set up svm params
	CvSVMParams params;
	params.svm_type = CvSVM::C_SVC;
	params.kernel_type = CvSVM::LINEAR;
	params.term_crit = cvTermCriteria(CV_TERMCRIT_ITER, 100, 1e-6);

	assign_lables();

	Mat labels_Mat(dataset_size, 1, CV_32FC1, labels);
	Mat training_Mat(dataset_size, number_of_features, CV_32FC1, data);

	// train the svm
	CvSVM SVM;
	SVM.train(training_Mat, labels_Mat, Mat(), Mat(), params);
	CvFileStorage* fs = cvOpenFileStorage("dataset.xml", 0, CV_STORAGE_WRITE);
	if (fs) {
		SVM.write(fs, "svm");
	}
	cvReleaseFileStorage(&fs);
}

void trainner::assign_lables() {
	int index = 0;
	////	// up
	//	for (int i = 0; i < 5; i++) {
	//		labels[index] = 1.0;
	//		index++;
	//	}
	//	//open
	//	for (int i = 0; i < 11; i++) {
	//		labels[index] = 2.0;
	//		index++;
	//	}
	//	// capture
	//	for (int i = 0; i < 19; i++) {
	//		labels[index] = 3.0;
	//		index++;
	//	}
	//	//call
	//	for (int i = 0; i < 12; i++) {
	//		labels[index] = 4.0;
	//		index++;
	//	}
	//	//left
	//	for (int i = 0; i < 11; i++) {
	//		labels[index] = 5.0;
	//		index++;
	//	}
	//	//right
	//	for (int i = 0; i < 11; i++) {
	//		labels[index] = 6.0;
	//		index++;
	//	}
	//	//closed
	//	for (int i = 0; i < 6; i++) {
	//		labels[index] = 7.0;
	//		index++;
	//	}
	//	//start
	//	for (int i = 0; i < 7; i++) {
	//		labels[index] = 8.0;
	//		index++;
	//	}
	// separate test
	//	//	// up
	//		for (int i = 0; i < 8; i++) {
	//			labels[index] = 1.0;
	//			index++;
	//		}
	//		//open
	//		for (int i = 0; i < 8; i++) {
	//			labels[index] = 2.0;
	//			index++;
	//		}
	//		// capture
	//		for (int i = 0; i < 10; i++) {
	//			labels[index] = 3.0;
	//			index++;
	//		}
	//		//call
	//		for (int i = 0; i < 12; i++) {
	//			labels[index] = 4.0;
	//			index++;
	//		}
	//		//left
	//		for (int i = 0; i < 9; i++) {
	//			labels[index] = 5.0;
	//			index++;
	//		}
	//		//right
	//		for (int i = 0; i < 6; i++) {
	//			labels[index] = 6.0;
	//			index++;
	//		}
	//		//closed
	//		for (int i = 0; i < 9; i++) {
	//			labels[index] = 7.0;
	//			index++;
	//		}
	//		//start
	//		for (int i = 0; i < 11; i++) {
	//			labels[index] = 8.0;
	//			index++;
	//		}
	//
	// after first combine
	// up
	for (int i = 0; i < /*14*/11; i++) {
		labels[index] = 1.0;
		index++;
	}
	//open
	for (int i = 0; i < /*22*/19; i++) {
		labels[index] = 2.0;
		index++;
	}
	// capture
	for (int i = 0; i < /*38*/30; i++) {
		labels[index] = 3.0;
		index++;
	}
	//call
	for (int i = 0; i < /*25*/23; i++) {
		labels[index] = 4.0;
		index++;
	}
	//left
	for (int i = 0; i < 18/*22*/; i++) {
		labels[index] = 5.0;
		index++;
	}
	//right
	for (int i = 0; i < /*30*/24; i++) {
		labels[index] = 6.0;
		index++;
	}
	//closed
	for (int i = 0; i < 11/*15*/; i++) {
		labels[index] = 7.0;
		index++;
	}
	//start
	for (int i = 0; i < 13/*18*/; i++) {
		labels[index] = 8.0;
		index++;
	}
	//Lup
	for (int i = 0; i < 10; i++) {
		labels[index] = 9;
		index++;
	}
	//Lopen
	for (int i = 0; i < 14/*15*/; i++) {
		labels[index] = 10;
		index++;
	}
	//Lcapture
	for (int i = 0; i < 17; i++) {
		labels[index] = 11;
		index++;
	}
	//Lcall
	for (int i = 0; i < 7; i++) {
		labels[index] = 12;
		index++;
	}
	//false
	for (int i = 0; i < 1; i++) {
		labels[index] = 13;
		index++;
	}
}
